share/spack/setup-env.fish

```diff
@@ -0,0 +1,826 @@
+            echo $elt
+        else
+            # bash compatibility: stop when the match first fails. Upon failure,
+            # we pack the remainder of `argv` into a global `__sp_remaining_args`
+            # array (`i` tracks the index of the next element).
+            set __sp_remaining_args (stream_args $argv[$i..-1])
+            return
+        end
+
+        # increment argument counter: used in place of bash's `shift` command
+        set -l i (math $i+1)
+
+    end
+
+    # if all elements in `argv` are matched, make sure that `__sp_remaining_args`
+    # is deleted (this might be overkill...).
+    set -e __sp_remaining_args
+end
+
+
+function get_mod_args -d "return submodule flags"
+    #
+    # Accumulate subcommand and submodule arguments. These are accumulated into
+    # the external arrays `__sp_subcommand_args`, and `__sp_module_args`. NOTE:
+    # Sets the external array: `__sp_remaining_args` containing all unprocessed
+    # arguments.
+    #
+
+    set __sp_subcommand_args
+    set __sp_module_args
+
+    # initialize argument counter
+    set -l i 1
+
+    for elt in $argv
+
+        if echo $elt | string match -r -q "^-"
+
+            # Notes: [2] (cf. EOF)
+            if test "x$elt" = "x-r"
+                set __sp_subcommand_args $__sp_subcommand_args $elt
+            else if test "x$elt" = "x--dependencies"
+                set __sp_subcommand_args $__sp_subcommand_args $elt
+            else
+                set __sp_module_args $__sp_module_args $elt
+            end
+
+        else
+            # bash compatibility: stop when the match first fails. Upon failure,
+            # we pack the remainder of `argv` into a global `__sp_remaining_args`
+            # array (`i` tracks the index of the next element).
+            set __sp_remaining_args (stream_args $argv[$i..-1])
+            return
+        end
+
+        # increment argument counter: used in place of bash's `shift` command
+        set -l i (math $i+1)
+
+    end
+
+    # if all elements in `argv` are matched, make sure that `__sp_remaining_args`
+    # is deleted (this might be overkill...).
+    set -e __sp_remaining_args
+end
+
+
+
+
+#
+# CHECK_SP_FLAGS, CONTAINS_HELP_FLAGS, CHECK_ENV_ACTIVATE_FLAGS, and 
+# CHECK_ENV_DEACTIVATE_FLAGS: support functions for checking arguments and flags.
+#
+
+function check_sp_flags -d "check spack flags for h/V flags"
+    #
+    # Check if inputs contain h or V flags.
+    #
+
+    # combine argument array into single string (space seperated), to be passed
+    # to regular expression matching (`string match -r`)
+    set -l _a "$argv"
+
+    # skip if called with blank input. Notes: [1] (cf. EOF)
+    if test -n "$_a"
+        if echo $_a | string match -r -q ".*h.*"
+            return 0
+        end
+        if echo $_a | string match -r -q ".*V.*"
+            return 0
+        end
+    end
+
+    return 1
+end
+
+
+function contains_help_flags -d "checks for help (-h/--help) flags"
```

Comment:
I think these flag checking functions will have to be updated like they are in #15429

Reply:
The way I understand https://github.com/spack/spack/pull/15429, we want to make sure that `contains_help_flags` only matches `-h` and `--help` and __not__ `asdf-h`. I added this messy-looking regex `([^\S]*-h|^[\s]*-h[\s]*\$)` to take care of that (of course I document it in the preceding comment)